README.md
<div align="center">

# **DeltaFrame**

<img width="250" height="250" src="https://raw.githubusercontent.com/robertcorponoi/graphics/main/deltaframe/deltaframe-logo.png">

DeltaFrame is a lightweight animation and game loop manager.

[![NPM version](https://img.shields.io/npm/v/deltaframe.svg?style=flat)](https://www.npmjs.com/package/deltaframe)
[![Known Vulnerabilities](https://snyk.io/test/github/robertcorponoi/deltaframe/badge.svg)](https://snyk.io/test/github/robertcorponoi/deltaframe)
![npm](https://img.shields.io/npm/dt/deltaframe)
<a href="https://badge.fury.io/js/deltaframe"><img src="https://img.shields.io/github/license/robertcorponoi/deltaframe.svg" alt="license" height="20"></a>

</div>

## **Installation**

```sh
npm install deltaframe
```

## **Usage**

After installing Deltaframe, it can be initialized like so:

```ts
const deltaframe = new Deltaframe();
```

Deltaframe also accepts an options object at initialization with the following options available:

| param              | type    | description                                                                                                                              | default  |
| ------------------ | ------- | ---------------------------------------------------------------------------------------------------------------------------------------- | -------- |
| minFps             | number  | The minimum number of frames per second to run deltaframe at. If frames per second drop below this, deltaframe will attempt to restart.  | 5        |
| targetFps          | number  | The number of frames per second that deltaframe should achieve.                                                                          | 60       |
| maxRestartAttempts | number  | The number of times deltaframe will attempt to restart before stopping entirely.                                                         | Infinity |
| runTime            | number  | Specify a value in milliseconds to have Deltaframe automatically stop after the specified amount of time has passed.                     | Infinity |
| forceSetTimeout    | boolean | Indicates whether setTimeout should be used instead of requestAnimationFrame even if requestAnimation is supported by the user's browser | false    |

So an example of initializating Deltaframe with options is:

```js
const options = {
    minFps: 40,
    maxRestartAttempts: 10,
};

const deltaframe = new Deltaframe(options);
```

Deltaframe all revolves around a function you pass to it and this function should contain your drawing code.

For example, we'll take a simple game loop that moves a circle.

Traditionally, with `requestAnimationFrame` it would be accomplished like so:

```js
const canvas = document.getElementById("myCanvas");
const ctx = canvas.getContext("2d");

let x = canvas.width / 2;
let y = canvas.height - 30;

function drawCircle() {
    ctx.beginPath();
    ctx.arc(x, y - 30, 10, 0, Math.PI * 2);
    ctx.fillStyle = "#0095DD";
    ctx.fill();
    ctx.closePath();
}

function draw(time) {
    ctx.clearRect(0, 0, canvas.width, canvas.height);

    drawCircle();

    x += 2;
    y += -2;

    requestAnimationFrame(draw);
}

requestAnimationFrame(draw);
```

This example uses `drawCircle` to just draw a circle centered horizontally and near the bottom of the canvas
and then it uses `requestAnimationFrame` recursively to keep the loop going and it makes the ball go diagonally.

To change this so that it uses deltaframe instead, use the following:

```js
const canvas = document.getElementById("myCanvas");
const ctx = canvas.getContext("2d");

let x = canvas.width / 2;
let y = canvas.height - 30;

function drawCircle() {
    ctx.beginPath();
    ctx.arc(x, y - 30, 10, 0, Math.PI * 2);
    ctx.fillStyle = "#0095DD";
    ctx.fill();
    ctx.closePath();
}

function draw(time, delta, deltaAverage) {
    ctx.clearRect(0, 0, canvas.width, canvas.height);

    drawCircle();

    x += 2;
    y += -2;
}

deltaframe.start(draw);
```

The biggest difference here is removing all calls to `requestAnimationFrame` and just using `deltaframe.start(draw)` instead. You'll also notice now that the draw function now accepts two more parameters because deltaframe returns three values which can be used inside of your drawing function. The `time` parameter is the same as it is in the first example, its the DOMHighResTimestamp as returned from `requestAnimationFrame` or (`window.performance.now()` in the case of `setTimeout`). The `delta` parameter is the change in time from the last frame to this frame and `deltaAverage` is the mean of the most recent 10 delta values. These values can all be used inside of the `draw` function however you deem fit.

## **API**

### **start**

The start method takes a function and begins running the deltaframe loop on it.

| param | type     | description                             | default |
| ----- | -------- | --------------------------------------- | ------- |
| fn    | Function | The function to call on a drawing loop. |         |

```js
const canvas = document.getElementById("myCanvas");
const ctx = canvas.getContext("2d");

let x = canvas.width / 2;
let y = canvas.height - 30;

function drawCircle() {
    ctx.beginPath();
    ctx.arc(x, y - 30, 10, 0, Math.PI * 2);
    ctx.fillStyle = "#0095DD";
    ctx.fill();
    ctx.closePath();
}

function draw(time, delta, deltaAverage) {
    ctx.clearRect(0, 0, canvas.width, canvas.height);

    drawCircle();

    x += 2;
    y += -2;
}

deltaframe.start(draw);
```

### **pause**

The pause method temprarily stops the deltaframe loop. This should only be used if you plan on resuming it at some point and not just stopping entirely.

Note that deltaframe automatically pauses when the user switches tabs.

```js
// Start the drawing loop.
deltaframe.start(draw);

setTimeout(() => {
    // After one second the drawing loop will pause and the game will look 'frozen'.
    deltaframe.pause();
}, 1000);
```

### **resume**

The resume method resumes the deltaframe loop after a paused state. Everything will be where it was before it was paused so it will seem like it was never
even paused.

Note that deltaframe automatically resumes from being paused when the user switches back to the tab containing deltaframe.

```js
// Start the drawing loop.
deltaframe.start(draw);

setTimeout(() => {
    // After one second the drawing loop will pause and the game will look 'frozen'.
    deltaframe.pause();

    setTimeout(() => {
        // After two seconds resume the drawing loop like it was never even paused.
        deltaframe.resume();
    }, 2000);
}, 1000);
```

### **stop**

The stop method stops the game and resets all deltaframe values back to their original values.

Note that this should only be used when you are done with the animation or drawing. In order to start the loop again you have to call start and provide the
drawing function.

```js
// Start the drawing loop.
deltaframe.start(draw);

// If some imaginary condition is met, stop deltaframe from running any more.
if (someConditionThatEndsTheAnimation) {
    deltaframe.stop();
}
```

## **License**

[MIT](./LICENSE)


CHANGELOG.md
## 3.0.0 / 2023-03-22

-   Changed the order of the parameters in the update function from (time, delta, deltaAverage) to (delta, deltaAverage, time) since most update loops only use delta.
-   Changed Options from a class to a type.
-   Updated various types.
-   Added ESLint and Prettier.
-   Added GitHub workflow for creating the bundle and running the tests on PRs.
-   Added dependabot config.
-   Updated tests to use jest which so they no longer have to run in a browser.
-   Updated out-of-date dependencies to their latest versions.

## 2.0.4 / 2020-04-17

-   [MISC] Updated out-of-date dependencies to their latest versions which also fixed all possible fixed security vulnerabilities.

## 2.0.3 / 2020-03-31

-   [MISC] Updated out-of-date dependencies to their latest versions.

## 2.0.2 / 2020-03-09

-   [MISC] Changed from using rollup-plugin-commonjs and rollup-plugin-node-resolve to using @rollup/plugin-commonjs and @rollup/plugin-node-resolve.
-   [MISC] Updated dependencies with security vulnerabilities.
-   [MISC] Updated out-of-date dependencies to their latest versions.

## 2.0.1 / 2020-02-18

-   [DOCS] Updated documentation to reflect new changes.

## 2.0.0 / 2020-02-18

-   [FEATURE] Separated extra 'bulky' features to deltaframe-extra. This package will be dedicated to just being a lightweight game loop runner while deltaframe-extra will consist of other feature such as tasks.
-   [FEATURE] Removed tasks from this and moved them to deltaframe-extra.
-   [MISC] Updated dependencies to their latest versions.
-   [MISC] Removed mocha and chai dependencies as they're referenced by script tags in the test index.html.
-   [MISC] Changed CHANGELOG format.

## 1.3.0 / 2020-01-23

-   [FEATURE] Added ability to add tasks that can run one or more times on an interval with an optional delay.
-   [DOCS] Added documentation about tasks.
-   [TEST] Added global to mocha setup to prevent an irrelevant error from displaying.

## 1.2.0 / 2020-01-14

-   [FEATURE] Added `time` as a getter.
-   [MISC] Removed unnecessary jsdoc comments.
-   [MISC] Removed `public` from methods as its redundant.

## 1.1.7 / 2020-01-14

-   [MISC] Updated dev dependencies to their latest versions.
-   [MISC] Updated license year to reflect new year.
-   [MISC] Moved logo to graphics repo and updated reference in the README.

## 1.1.6 / 2019-12-26

-   [MISC] Updated dev dependencies to their latest versions.
-   [MISC] Added testing section in README.
-   [MISC] Cleaned up spacinig in code blocks in README.

## 1.1.4 / 2019-12-08

-   [MISC] Updated dev dependencies to their latest versions.

## 1.1.3 / 2019-11-03

-   [FEATURE] Removed testing dependencies and replaced them with unpkg scripts.
-   [MISC] Added more badges to README.

## 1.1.1 / 2019-11-03

-   [FEATURE] Updated all dependencies to their latest versions.

## 1.1.0 / 2019-07-29

-   [FEATURE] Added the visibility change functionality fully so that the timer pauses and resumes depending on whether or not the page is in focus.
-   [FEATURE] Cleaned up tests of unnecessary comments.

## 1.0.5 / 2019-07-29

-   [FEATURE] Updated all dev dependencies to their latest versions.

## 1.0.4 / 2019-07-17

-   [HOTFIX] Updated all dev dependencies to their latest versions and fixed any security vulnerabilities found in them.

## 1.0.3 / 2019-05-12

-   Made the initialization options optional.
-   Updated Typescript definitions.

## 1.0.2 / 2019-05-12

-   Cleaned up unused options from tsconfig
-   Updated all dev dependencies to the latest versions.
-   Updated documentation.

## 1.0.1 / 2019-03-10

-   Added missing `typings` key in `package.json`.
-   Moved the location of the bundled file from `dist/deltaframe.js` to the root directory.
-   Updated all dev dependencies.

## 1.0.0 / 2018-01-26

-   Added Typescript definitions with Babel transpiling.
-   Added rollup as the module bundler which exports to `src/deltaframe.js`.
-   Changed the initialization options into their own Object for greater flexibility.
-   The `minFps` and `targetFps` are now automatically converted to decimals in the new `Options` object.
-   Deltaframe is now exported as a default module so instead of using `import { Deltaframe } from './path/to/deltaframe.js` you can now use `import Deltaframe from './path/to/deltaframe.js'`.

## 0.2.0 / 2018-12-20

-   Added getters for various properties:
    -   `timesRestarted`: Returns the number of times Deltaframe has had to restart due to the average fps dipping below the minimum fps.
    -   `isRunning`: Returns `true` if the animation loop is currently running.
    -   `isPaused`: Returns `true` if Deltaframe is paused.
    -   `frame`: Returns the current frame number that animation is on.
    -   `time`: Returns the current time of the animation.
    -   `delta`: Returns the average time between frames at that moment.
-   Changed the visibility change event to use `addEventListener` instead of `onVisibilityChange`.
-   Added a new initialization option called `runTime` which can be used to specify a time, in milliseconds, that Deltaframe should run the animation for.
-   Expanded test cases to account for the new `runTime` option being added.

## 0.1.0 / 2018-08-26

-   Initial Release


deltaframe.js
function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

var Options = /*#__PURE__*/function () {
  /**
   * The lowest the fps can drop to before the Deltaframe restarts to attempt to fix the problem.
   * 
   * @property {number}
    * 
    * @default 15
   */

  /**
   * The fps that the game loop should aim to  achieve.
   * 
   * @property {number}
    * 
    * @default 60
   */

  /**
   * When the fps goes below the minFps Deltaframe will restart. This indicates how many times it will  restart before stopping permanently.
   * 
   * @property {number}
    * 
    * @default Infinity
   */

  /**
   * Specify the amount of milliseconds that Deltaframe should run for.
   * 
   * @property {number}
    * 
    * @default Infinity
   */

  /**
   * Indicates whether setTimeout should be used even if requestAnimationFrame is supported by the user's browser.
   * 
   * @property {number}
    * 
    * @default false
   */

  /**
    * @param {Object} options The initialization options passed to Deltaframe.
    */
  function Options() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, Options);

    _defineProperty(this, "minFps", 15);

    _defineProperty(this, "targetFps", 60);

    _defineProperty(this, "maxRestartAttempts", Infinity);

    _defineProperty(this, "runTime", Infinity);

    _defineProperty(this, "forceSetTimeout", false);

    Object.assign(this, options);
  }
  /**
   * Return the minFps as a decimal representing the amount of time before a frame should occur.
   * 
   * @returns {number}
   */


  _createClass(Options, [{
    key: "minFpsCalc",
    get: function get() {
      return Math.floor(1000 / this.minFps);
    }
    /**
     * Return the targetFps as a decimal representing the amount of time before a frame should occur.
     * 
     * @returns {number}
     */

  }, {
    key: "targetFpsCalc",
    get: function get() {
      return Math.floor(1000 / this.targetFps);
    }
  }]);

  return Options;
}();

var RequestAnimationFrame = /*#__PURE__*/function () {
  /**
   * A reference to the id returned by requestAnimationFrame or setTimeout so  that we can cancel their operation when needed.
   * 
   * @property {number}
   */

  /**
   * Keeps track of whether the loop is already running or not so it's not accidently restarted.
   * 
   * @property {boolean}
   * 
   * @default false
   */

  /**
   * The function that should be run on every update of the loop.
   * 
   * @property {Function}
   * 
   * @default ()=>{}
   */

  /**
   * Indicates whether setTImeout is being used instead of requestAnimationFrame.
   * 
   * @property {boolean}
   * 
   * @default false
   */
  function RequestAnimationFrame() {
    _classCallCheck(this, RequestAnimationFrame);

    _defineProperty(this, "id", 0);

    _defineProperty(this, "running", false);

    _defineProperty(this, "fn", function () {});

    _defineProperty(this, "usingSetTimeout", false);

    /**
     * Use the version of requestAnimationFrame that is supported by the user's browser and if none are  supported, use setTimeout instead.
     * 
     * @property {RequestAnimationFrame|setTimeout}
     */
    window.requestAnimationFrame = window.requestAnimationFrame || window.webkitRequestAnimationFrame || function (f) {
      return setTimeout(f, 1000 / 60);
    };
    /**
     * Use the version of cancelAnimationFrame that is supported by the user's browser and if none are supported,  then setTimeout was used 
     * and so we use clearTimeout instead.
     * 
     * @property {cancelAnimationFrame}
     */


    window.cancelAnimationFrame = window.cancelAnimationFrame || window.webkitCancelAnimationFrame || function () {
      clearTimeout(this.id);
    };
  }
  /**
   * Start the operation of the requestAnimationFrame or setTimeout loop.
   * 
   * @param {Function} fn The function to run every update of the loop.
   * @param {boolean} forceSetTimeout Indicates whether setTimeout should be used even if the user's browser supports requestAnimationFrame.
   */


  _createClass(RequestAnimationFrame, [{
    key: "start",
    value: function start(fn, forceSetTimeout) {
      var _this = this;

      if (this.running) return;
      this.running = true;
      this.fn = fn;

      if (forceSetTimeout) {
        this.usingSetTimeout = true;
        this.updateTimeout();
      } else {
        window.requestAnimationFrame(function (time) {
          return _this.updateRAF(time);
        });
      }
    }
    /**
     * Call requestAnimationFrame recursively so that the loop keeps going and also send the timestamps over to Deltaframe.
     * 
     * @param {number} timestamp The timestamp from the most recent requestAnimationFrame request.
     */

  }, {
    key: "updateRAF",
    value: function updateRAF(timestamp) {
      var _this2 = this;

      this.running = true;
      this.fn(timestamp);
      this.id = window.requestAnimationFrame(function (time) {
        return _this2.updateRAF(time);
      });
    }
    /**
     * Call setTimeout recursively so that the loop keeps going and also send the timestamps over to Deltaframe.
     */

  }, {
    key: "updateTimeout",
    value: function updateTimeout() {
      var _this3 = this;

      var timestamp = window.performance.now();
      this.fn(timestamp);
      this.id = window.setTimeout(function () {
        return _this3.updateTimeout();
      }, 1000 / 60);
    }
    /**
     * Restart the requestAnimation or setTimeout loop.
     */

  }, {
    key: "restart",
    value: function restart() {
      var _this4 = this;

      if (this.usingSetTimeout) window.clearTimeout(this.id);else window.cancelAnimationFrame(this.id);
      this.id = 0;
      this.running = false;
      if (this.usingSetTimeout) this.updateTimeout();else window.requestAnimationFrame(function (time) {
        return _this4.updateRAF(time);
      });
      this.running = true;
    }
    /**
     * Stop the loop by calling cancelAnimationFrame or clearTimeout.
     */

  }, {
    key: "stop",
    value: function stop() {
      if (this.usingSetTimeout) window.clearTimeout(this.id);else window.cancelAnimationFrame(this.id);
      this.id = 0;
      this.running = false;

      this.fn = function () {};

      return;
    }
  }]);

  return RequestAnimationFrame;
}();

/**
 * Deltaframe is an animation and game loop manager that makes sure your application is punctual and performant.
 */

var Deltaframe = /*#__PURE__*/function () {
  /**
   * A reference to the options for this instance of Deltaframe.
   * 
   * @private
   * 
   * @property {Options}
   */

  /**
   * The amount of times Deltaframe has had to restart due to the average fps dipping below the minimum fps for a 
   * series of frames.
   * 
   * @private
   * 
   * @property {number}
   */

  /**
   * Indicates whether Deltaframe is currently is currently running and not pausedor stopped.
   * 
   * @private
   * 
   * @property {boolean}
   */

  /**
   * Indicates whether Deltaframe is currently paused.
   * 
   * @private
   * 
   * @property {boolean}
   */

  /**
   * The function that will be called on every Deltaframe update.
   * 
   * @private
   * 
   * @property {Function}
   */

  /**
   * The current frame that Deltaframe is on.
   * 
   * @private
   * 
   * @property {number}
   */

  /**
   * The current timestamp as of the latest call to RequestAnimationFrame.
   * 
   * @private
   * 
   * @property {DOMHighResTimeStamp|number}
   */

  /**
   * The timestamp before the current timestamp.
   * 
   * @private
   * 
   * @property {DOMHighResTimeStamp|number}
   */

  /**
   * The difference in time between the current time and the last time.
   * 
   * @private
   * 
   * @property {number}
   */

  /**
   * The average difference in time between frames.
   * 
   * @private
   * 
   * @property {number}
   */

  /**
   * A set of up to 10 recent previous delta values that are used to get the mean delta.
   * 
   * @private
   * 
   * @property {Array<number>}
   */

  /**
   * Since we only want to go up to 10 on the deltaHistory, we keep track of what index we're  on so we can reset to 0 once were at 10.
   * 
   * @private
   * 
   * @property {number}
   */

  /**
   * Initialize the RequestAnimationFrame abstraction module.
   * 
   * @private
   * 
   * @property {RequestAnimationFrame}
   */

  /**
   * Use the version of hidden that's supported by the user's browser.
   * 
   * @private
   * 
   * @property {document.hidden}
   */

  /**
   * @param {Object} [options] The options to pass to this Deltaframe instance.
   * @param {number} [options.minFps=15] The minimum fps value allowed before Deltaframe will restart to try to correct the issue.
   * @param {number} [options.targetFps=60] The fps that Deltaframe should aim to achieve.
   * @param {number} [options.maxRestartAttempts=Infinity] The number of times Deltaframe will restart due to problems before stopping entirely.
   * @param {number} [options.runTime=Infinity] The length of time that this instance of Deltaframe will run. This can be used to create an animation that lasts a specific amount of time.
   * @param {boolean} [options.forceSetTimeout=false] If set to true, Deltaframe will use setTimeout for the loop instead of requestAnimationFrame.
   */
  function Deltaframe(options) {
    _classCallCheck(this, Deltaframe);

    _defineProperty(this, "_options", void 0);

    _defineProperty(this, "_restartAttempts", void 0);

    _defineProperty(this, "_running", void 0);

    _defineProperty(this, "_paused", void 0);

    _defineProperty(this, "_fn", void 0);

    _defineProperty(this, "_frame", void 0);

    _defineProperty(this, "_time", void 0);

    _defineProperty(this, "_prevTime", void 0);

    _defineProperty(this, "_delta", void 0);

    _defineProperty(this, "_deltaAverage", void 0);

    _defineProperty(this, "_deltaHistory", void 0);

    _defineProperty(this, "_deltaIndex", void 0);

    _defineProperty(this, "_raf", void 0);

    _defineProperty(this, "_hidden", void 0);

    this._options = new Options(options);
    this._restartAttempts = 0;
    this._running = false;
    this._paused = false;

    this._fn = function () {};

    this._frame = 0;
    this._time = 0;
    this._prevTime = 0;
    this._delta = 0;
    this._deltaAverage = 0;
    this._deltaHistory = [];
    this._deltaIndex = 0;
    this._raf = new RequestAnimationFrame();
    this._hidden = document.hidden;

    this._boot();
  }
  /**
   * Return the number of times that Deltafram has restarted.
   * 
   * @returns {number}
   */


  _createClass(Deltaframe, [{
    key: "start",

    /**
     * Start the loop.
     * 
     * @param {Function} fn The function to be called every step by the loop.
     */
    value: function start(fn) {
      var _this = this;

      this._fn = fn;
      this._prevTime = 0;
      this._running = true;

      this._raf.start(function (timestamp) {
        return _this._update(timestamp);
      }, this._options.forceSetTimeout);
    }
    /**
     * Pause the loop operation saving the state to be resumed at a later time.
     */

  }, {
    key: "pause",
    value: function pause() {
      this._paused = true;
      this._running = false;
    }
    /**
     * Resume the loop from a paused state.
     */

  }, {
    key: "resume",
    value: function resume() {
      this._paused = false;
      this._prevTime = window.performance.now();
      this._running = true;
    }
    /**
     * Stop the loop and reset all time values of Deltaframe.
     */

  }, {
    key: "stop",
    value: function stop() {
      var _this2 = this;

      this._restartAttempts = 0;
      this._running = false;
      this._paused = false;

      this._fn = function () {};

      this._frame = 0;
      this._time = 0;
      this._prevTime = 0;
      this._delta = 0;
      this._deltaHistory = [];
      this._deltaIndex = 0;
      document.removeEventListener('visibilitychange', function () {
        return _this2._visibilityChange;
      });

      this._raf.stop();

      return;
    }
    /**
     * Initialize the page visibility events which will let us save resources by pausing our updates when the user is not 
     * interacting with the page running Deltaframe.
     * 
     * @private
     */

  }, {
    key: "_boot",
    value: function _boot() {
      var _this3 = this;

      document.addEventListener("visibilitychange", function () {
        return _this3._visibilityChange();
      });
    }
    /**
     * Update is called whenever requestAnimationFrame decides it can process the next step of the loop  or roughly 60 
     * times per second using setTimeout.
     * 
     * @private
     * 
     * @param {DOMHighResTimeStamp|number} timestamp The timestamp as returned from requestAnimationFrame.
     */

  }, {
    key: "_update",
    value: function _update(timestamp) {
      if (this._paused) return;

      if (timestamp >= this._options.runTime) {
        this.stop();
        return;
      }

      this._time = timestamp;
      this._delta = timestamp - this._prevTime;
      if (this._deltaIndex === 10) this._deltaIndex = 0;
      this._deltaHistory[this._deltaIndex] = this._delta;
      this._deltaIndex++;
      var mean = 0;

      for (var i = 0; i < this._deltaHistory.length; ++i) {
        mean += this._deltaHistory[i];
      }

      mean /= 10;
      this._deltaAverage = mean;

      if (this._deltaAverage >= this._options.minFpsCalc) {
        if (this._restartAttempts === this._options.maxRestartAttempts) {
          this.stop();
          return;
        }

        this._raf.restart();

        this._restartAttempts++;
      }

      if (this._deltaAverage >= this._options.targetFpsCalc) {
        this._frame++;

        this._fn(timestamp, this._delta, this._deltaAverage);

        this._prevTime = timestamp;
      }
    }
    /**
     * When the the user has switched to a different tab and is not on the same page that Deltaframe is running on, Deltaframe 
     * will pause and when the user comes back it will resume.
     * 
     * @private
     */

  }, {
    key: "_visibilityChange",
    value: function _visibilityChange() {
      var visibility = document.visibilityState;
      if (this.isPaused && visibility === 'visible') this.resume();else if (this.isRunning && visibility === 'hidden') this.pause();
    }
  }, {
    key: "timesRestarted",
    get: function get() {
      return this._restartAttempts;
    }
    /**
     * Returns if Deltaframe is running or not.
     * 
     * @returns {boolean}
     */

  }, {
    key: "isRunning",
    get: function get() {
      return this._running;
    }
    /**
     * Returns if Deltaframe is paused or not.
     * 
     * @returns {boolean}
     */

  }, {
    key: "isPaused",
    get: function get() {
      return this._paused;
    }
    /**
     * Returns the current frame.
     * 
     * @returns {number}
     */

  }, {
    key: "frame",
    get: function get() {
      return this._frame;
    }
    /**
     * Returns the current time.
     * 
     * @returns {DOMHighResTimeStamp|number}
     */

  }, {
    key: "time",
    get: function get() {
      return this._time;
    }
  }]);

  return Deltaframe;
}();

export default Deltaframe;
